packages/core/src/render3/debug.ts

```diff
@@ -0,0 +1,230 @@
+ *       {html: '<span>', ..., nodes: null}
+ *     ]}
+ *   ]
+ * }
+ * ```
+ */
+
+
+export function attachLViewDebug(lView: LView) {
+  (lView as any).debug = new LViewDebug(lView);
+}
+
+export function attachLContainerDebug(lContainer: LContainer) {
+  (lContainer as any).debug = new LContainerDebug(lContainer);
+}
+
+export function toDebug(obj: LView): LViewDebug;
+export function toDebug(obj: LView | null): LViewDebug|null;
+export function toDebug(obj: LView | LContainer | null): LViewDebug|LContainerDebug|null;
+export function toDebug(obj: any): any {
+  if (obj) {
+    const debug = (obj as any).debug;
+    assertDefined(debug, 'Object does not have a debug representation.');
+    return debug;
+  } else {
+    return obj;
+  }
+}
+
+/**
+ * Use this method to unwrap a native element in `LView` and convert it into HTML for easier
+ * reading.
+ *
+ * @param value possibly wrapped native DOM node.
+ * @param includeChildren If `true` than the serialized HTML form will include child elements (same
+ * as `outerHTML`). If `false` than the serialized HTML form will only contain the element itself
+ * (will not serialize child elements).
+ */
+function toHtml(value: any, includeChildren: boolean = false): string|null {
+  const node: HTMLElement|null = readElementValue(value) as any;
+  if (node) {
+    const outerHTML = node.outerHTML;
+    const innerHTML = node.innerHTML;
+    if (includeChildren || innerHTML == null) {
+      return outerHTML;
+    } else {
+      return outerHTML.replace('>' + innerHTML, '>ɵ').split('ɵ').shift() || null;
+    }
+  } else {
+    return null;
+  }
+}
+
+export class LViewDebug {
+  constructor(private readonly _raw_lView: LView) {}
+
+  /**
+   * Flags associated with the `LView` unpacked into a more readable state.
+   */
+  get flags() {
+    const flags = this._raw_lView[FLAGS];
+    return {
+      __raw__flags__: flags,
+      initPhaseState: flags & LViewFlags.InitPhaseStateMask,
+      creationMode: !!(flags & LViewFlags.CreationMode),
+      firstViewPass: !!(flags & LViewFlags.FirstLViewPass),
+      checkAlways: !!(flags & LViewFlags.CheckAlways),
+      dirty: !!(flags & LViewFlags.Dirty),
+      attached: !!(flags & LViewFlags.Attached),
+      destroyed: !!(flags & LViewFlags.Destroyed),
+      isRoot: !!(flags & LViewFlags.IsRoot),
+      indexWithinInitPhase: flags >> LViewFlags.IndexWithinInitPhaseShift,
+    };
+  }
+  get parent(): LViewDebug|LContainerDebug|null { return toDebug(this._raw_lView[PARENT]); }
+  get host(): string|null { return toHtml(this._raw_lView[HOST], true); }
+  get context(): {}|null { return this._raw_lView[CONTEXT]; }
+  /**
+   * The tree of nodes associated with the current `LView`. The nodes have been normalized into a
+   * tree structure with relevant details pulled out for readability.
+   */
+  get nodes(): DebugNode[]|null {
+    const lView = this._raw_lView;
+    const tNode = this.__other__.tView.firstChild;
```

Comment:
Why is this access through `__other__`, that requires accessing almost all of LView now. I know this is debug so perf is not really a concern, however I don't see why this pattern was chosen. Could we do:

```suggestion
    const tNode = this._raw_lView[TVIEW].firstChild;
```

There's some more occurrences in `LViewDebug.childViews` that accesses `__other__` in a loop.

Reply:
no particular reason. Changed.